runner.py

```diff
@@ -0,0 +1,653 @@
+                return False
+
+        res = False
+        try:
+            res = self.harness(proc)
+        except pexpect.exceptions.TIMEOUT as exc:
+            logging.debug(str(exc))
+            self.status = TestCase.FAIL_TIMEOUT
+        except pexpect.exceptions.EOF as exc:
+            logging.warning(f"Test {self.name} reached EOF")
+            logging.warning(str(exc))
+        except Exception as exc:
+            logging.warning(f"Test {self.name} exception:\n {exc}")
+            raise exc
+
+        if res:
+            self.status = TestCase.PASS
+
+        return res
+
+
+class TestCaseCustomHarness(TestCase):
+    def __init__(self, name, target, harness_path, syspage=None):
+        super().__init__(name, target, syspage)
+        self.load_module(harness_path)
+
+    def load_module(self, path):
+        self.spec = importlib.util.spec_from_file_location('harness', path.absolute())
+        self.test_module = importlib.util.module_from_spec(self.spec)
+        self.spec.loader.exec_module(self.test_module)
+
+        if hasattr(self.test_module, 'harness'):
+            self.harness = self.test_module.harness
+        else:
+            raise ValueError(f"harness function has not been found in {path}")
+
+
+class TestCaseUnit(TestCase):
+    def __init__(self, name, target, syspage):
+        super().__init__(name, target, syspage)
+        self.harness = UnitTestHarness.harness
+        self.unit_test_results = []
+
+    def report(self):
+        super().report()
+
+        for test in self.unit_test_results:
+            if test.status == UnitTest.FAIL:
+                logging.info('\t' + str(test))
+            else:
+                logging.debug('\t' + str(test))
+
+    def handle(self, proc):
+        res = super().handle(proc)
+
+        if self.status == TestCase.PASS:
+            self.unit_test_results = res
+
+        for test in self.unit_test_results:
+            if test.status == UnitTest.FAIL:
+                self.status = TestCase.FAIL
+                break
+
+        return res
+
+
+class TestCaseFactory:
+    @staticmethod
+    def create(test):
+        if test['type'] == 'unit':
+            return TestCaseUnit(
+                    name=test['name'],
+                    target=test['target'],
+                    syspage=test['syspage']
+            )
+        if test['type'] == 'harness':
+            return TestCaseCustomHarness(
+                    name=test['name'],
+                    target=test['target'],
+                    harness_path=test['harness'],
+                    syspage=test['syspage']
+            )
+
+        raise ValueError(f"Unknown TestCase type: {test}")
+
+
+class TargetBuilder:
+    SYSPAGE = {
+        'ia32-generic': [
+            'uart16550',
+            'pc-ata',
+            'psh'
+        ],
+        'riscv64-virt': [
+            'uart16550',
+            'dummyfs',
+            'psh'
+        ]
+    }
```

Comment:
I suggest making this `class` more abstract.
Here define `SYSPAGE` e.g empty `SYSPAGE = []`.
Specific `SYSPAGE` values should be defined in classes for particular targets e.g. for `ia32-generic`
```
class IA32Builder(TargetBuilder):
    SYSPAGE = ( 'uart16550', 'pc-ata', 'psh' )

    def __init__(self):
        super().__init__('ia32-generic')
```